gallery/v2_transforms/plot_transforms_v2.py

```diff
@@ -7,103 +7,253 @@
+#
+# What do I pass as input?
+# ------------------------
+#
+# Above, we've seen two examples: one where we passed a single image as input
+# i.e. ``out = transforms(img)``, and one where we passed both an image and
+# bounding boxes, i.e. ``out_img, out_bboxes = transforms(img, bboxes)``.
+#
+# In fact, transforms support **arbitrary input structures**. The input can be a
+# single image, a tuple, an arbitrarily nested dictionary... pretty much
+# anything. The same structure will be returned as output. Below, we use the
+# same detection transforms, but pass a tuple (image, target_dict) as input and
+# we're getting the same structure as output:
 
-# passing a plain tensor together with an explicit image, will not transform the former
-plain_tensor_image, image = transform(plain_tensor_image, image)
+target = {
+    "bboxes": bboxes,
+    "labels": torch.arange(bboxes.shape[0]),
+    "this_is_ignored": ("arbitrary", {"structure": "!"})
+}
 
-print(image.shape, plain_tensor_image.shape)
+# Re-using the transforms and definitions from above.
+out_img, out_target = transforms(img, target)
 
-# passing a plain tensor without an explicit image, will transform the former
-plain_tensor_image, _ = transform(plain_tensor_image, bounding_boxes)
+plot([(img, target["bboxes"]), (out_img, out_target["bboxes"])])
+print(f"{out_target['this_is_ignored']}")
 
-print(image.shape, plain_tensor_image.shape)
+# %%
+# We passed a tuple so we get a tuple back, and the second element is the
+# tranformed target dict. Transforms don't really care about the structure of
+# the input; as mentioned above, they only care about the **type** of the
+# objects and transforms them accordingly.
+#
+# *Foreign* objects like strings or ints are simply passed-through. This can be
+# useful e.g. if you want to associate a path with every single sample when
+# debugging!
+#
+# .. _passthrough_heuristic:
+#
+# .. note::
+#
+#     **Disclaimer** This note is slightly advanced and can be safely skipped on
+#     a first read.
+#
+#     Pure :class:`torch.Tensor` objects are, in general, treated as images (or
+#     as videos for video-specific transforms). Indeed, you may have noticed
+#     that in the code above we haven't used the
+#     :class:`~torchvision.datapoints.Image` class at all, and yet our images
+#     got transformed properly. Transforms follow the following logic to
+#     determine whether a pure Tensor should be treated as an image (or video),
+#     or just ignored:
+#
+#     * If we there is an :class:`~torchvision.datapoints.Image`,
+#       :class:`~torchvision.datapoints.Video`,
+#       or :class:`PIL.Image.Image` instance in the input, all other pure
+#       tensors are passed-through.
+#     * If there is no :class:`~torchvision.datapoints.Image` or
+#       :class:`~torchvision.datapoints.Video` instance, only the first pure
+#       :class:`torch.Tensor` will be transformed as image or video, while all
+#       others will be passed-through.
```

Comment:
Since this is an advanced note anyway, should we explain what "first" means in the context of arbitrary input structure? Basically, the first pure tensor that we find we recursing depth-wise through the input.